File: src/edu/stmc/HypTest.java
package edu.stmc;

public interface HypTest extends  Cloneable {

  /** Reset the state (can be used to restart the test) */
  void reset();

  /**
   * Update the test by adding an observation
   * @param passed Whether or not the sample satisfied the test or not.
   */
  void update(boolean passed);

  /**
   * Update the test by adding an observation
   * @param passed Number of samples that are passed
   * @param failed Number of samples that are failed
   * @requires {@code passed >= 0}
   * @requires {@code failed >= 0}
   */
  void update(int passed, int failed);

  /** Whether or not an answer has been already found */
  boolean shouldStopNow();

  /**
   * After {@link #shouldStopNow} returns true, this method is used to see if the result is known
   * @requires {@link #shouldStopNow} should return {@literal true}
   */
  boolean hasAnswer();

  /**
   * After {@link #shouldStopNow} and {@link #hasAnswer} both returned true, this is used to see if the null
   * hypothesis is rejected.
   * @requires {@link #shouldStopNow} should return {@literal true}
   * @requires {@link #hasAnswer} should return {@literal true}
   */
  boolean isNullRejected();

  /** Get the parameters of the simulation (including the computed one) as a string. */
  String getParametersString();

  HypTest cloneCopy();
}

File: src/edu/stmc_ORG/HypTest.scala
package edu.stmc

/** Hypothesis Testing */
trait HypTest extends Cloneable {

  /** Reset the test (can be used to restart the test). */
  def reset(): Unit

  /** Update the test by adding an observation.
    *
    * @param positive Whether or not the new sample is positive.
    * @note Implementations might add a constraint on the total number of samples. */
  def update(positive: Boolean)

  /** Update the test by adding multiple observations.
    *
    * @param positive Number of samples that are positive.
    * @param negative Number of samples that are negative.
    * @note Requires `positive >= 0` and `negative >= 0`.
    * @note Implementations might add a constraint on the total number of samples. */
  def update(positive: Int, negative: Int)

  /** Whether or not the test is completed. */
  def completed: Boolean

  /** Whether or not a decision is made (it is possible that a test terminates without choosing between the null and
    * alternative hypotheses).
    *
    * @note Requires [[completed]] to be `true`. */
  def decided: Boolean

  /** Whether the test rejects the null hypothesis in favor of the alternative hypothesis.
    *
    * @note Requires [[completed]] to be `true`. */
  def rejected: Boolean

  /** Whether the test is failed to reject the null hypothesis in favor of the alternative hypothesis.
    *
    * @note Requires [[completed]] to be `true`. */
  def failed_to_reject: Boolean

  /** Get parameters of the test (including the computed ones) as a string. */
  def parametersStr: String

  /** Create a copy of this instance, so changes in one does not affect changes in the other. */
  def copy(): HypTest
}
